### PR TITLE
[SPARK-9826][Core] Fix cannot use custom classes in log4j.properties

### PR DESCRIPTION
**Pull request by michellemay**
Refactor Utils class and create ShutdownHookManager.  

NOTE: Wasn't able to run /dev/run-tests on windows machine.
Manual tests were conducted locally using custom log4j.properties file with Redis appender and logstash formatter (bundled in the fat-jar submitted to spark)

ex:
log4j.rootCategory=WARN,console,redis
log4j.appender.console=org.apache.log4j.ConsoleAppender
log4j.appender.console.target=System.err
log4j.appender.console.layout=org.apache.log4j.PatternLayout
log4j.appender.console.layout.ConversionPattern=%d{yy/MM/dd HH:mm:ss} %p %c{1}: %m%n

log4j.logger.org.eclipse.jetty=WARN
log4j.logger.org.eclipse.jetty.util.component.AbstractLifeCycle=ERROR
log4j.logger.org.apache.spark.repl.SparkIMain$exprTyper=INFO
log4j.logger.org.apache.spark.repl.SparkILoop$SparkILoopInterpreter=INFO
log4j.logger.org.apache.spark.graphx.Pregel=INFO

log4j.appender.redis=com.ryantenney.log4j.FailoverRedisAppender
log4j.appender.redis.endpoints=hostname:port
log4j.appender.redis.key=mykey
log4j.appender.redis.alwaysBatch=false
log4j.appender.redis.layout=net.logstash.log4j.JSONEventLayoutV1

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 1]
```diff
@@ -559,7 +559,7 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
     // Make sure the context is stopped if the user forgets about it. This avoids leaving
     // unfinished event logs around after the JVM exits cleanly. It doesn't help if the JVM
     // is killed, though.
-    _shutdownHookRef = Utils.addShutdownHook(Utils.SPARK_CONTEXT_SHUTDOWN_PRIORITY) { () =>
+    _shutdownHookRef = ShutdownHookManager.addShutdownHook(ShutdownHookManager.SPARK_CONTEXT_SHUTDOWN_PRIORITY) { () =>
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: This line is too long and will fail style checks.

File: core/src/main/scala/org/apache/spark/util/ShutdownHookManager.scala [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+package org.apache.spark.util
+
+import java.io._
+import java.util.PriorityQueue
+
+import scala.util.{Failure, Success, Try}
+
+import org.apache.hadoop.fs.FileSystem
+
+import tachyon.client.TachyonFile
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/util/ShutdownHookManager.scala]: nit: should be together with `org.apache` import above.

File: core/src/main/scala/org/apache/spark/util/ShutdownHookManager.scala [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.util
+
+import java.io._
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/util/ShutdownHookManager.scala]: Seems like this could be just `java.io.File`?

**Comment by vanzin** [comment]
The original reason (from what I remember) to hide this under the Utils class was to avoid exposing the `ShutdownHookManager`, which clashes with a Hadoop class with the same name (different package though). I guess since Spark code should not use the Hadoop one, it's ok to use the same name.

The change looks ok, but let's see what tests say.

**Comment by vanzin** [comment]
ok to test

**Comment by vanzin** [comment]
Jenkins, test this please.

**Comment by vanzin** [comment]
ok to test

File: core/src/main/scala/org/apache/spark/SparkContext.scala (hunk 1 above)
Comment (michellemay) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: How do you suggest to split this line up?  I'm not too much familiar with scala best practices and style.

**Comment by michellemay** [comment]
Where do we see test results ?

**Comment by vanzin** [comment]
@michellemay the last build has a link to the results (it's failing the style checks).

**Comment by michellemay** [comment]
@vanzin about your comment on ShutdownHookManager, should I rename it?  

Also, while refactoring calls to Utils.addShutdownHook, I've seen instances of calls to Runtime.getRuntime.addShutdownHook..  Should I replace them with calls to the new class?

org\apache\spark\deploy\ExternalShuffleService.scala:126
org\apache\spark\deploy\mesos\MesosClusterDispatcher.scala:113
org\apache\spark\storage\ExternalBlockStore.scala:181

**Comment by vanzin** [comment]
> about your comment on ShutdownHookManager, should I rename it?

No, the name is fine. I was just giving some historical context.

> I've seen instances of calls to Runtime.getRuntime.addShutdownHook..

From a quick look, it seems that at least the one in `ExternalBlockStore` should be changed; the other two seem to be separate processes (i.e. they are not part of a Spark application) so that's questionable. But I'd say leave it for a different change (maybe file a separate bug so it doesn't fall through the cracks).

File: core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala [hunk 4]
```diff
@@ -30,7 +30,7 @@ import org.apache.spark.status.api.v1.{ApiRootResource, ApplicationInfo, Applica
   UIRoot}
 import org.apache.spark.ui.{SparkUI, UIUtils, WebUI}
 import org.apache.spark.ui.JettyUtils._
-import org.apache.spark.util.{SignalLogger, Utils}
+import org.apache.spark.util.{SignalLogger, Utils, ShutdownHookManager}
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala]: nit: sorting order

File: core/src/main/scala/org/apache/spark/deploy/worker/ExecutorRunner.scala [hunk 5]
```diff
@@ -28,7 +28,7 @@ import org.apache.spark.rpc.RpcEndpointRef
 import org.apache.spark.{SecurityManager, SparkConf, Logging}
 import org.apache.spark.deploy.{ApplicationDescription, ExecutorState}
 import org.apache.spark.deploy.DeployMessages.ExecutorStateChanged
-import org.apache.spark.util.Utils
+import org.apache.spark.util.{Utils, ShutdownHookManager}
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/deploy/worker/ExecutorRunner.scala]: nit: sorting order

File: core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala [hunk 6]
```diff
@@ -44,7 +44,7 @@ import org.apache.spark.broadcast.Broadcast
 import org.apache.spark.deploy.SparkHadoopUtil
 import org.apache.spark.executor.DataReadMethod
 import org.apache.spark.rdd.HadoopRDD.HadoopMapPartitionsWithSplitRDD
-import org.apache.spark.util.{SerializableConfiguration, NextIterator, Utils}
+import org.apache.spark.util.{SerializableConfiguration, NextIterator, Utils, ShutdownHookManager}
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala]: nit: sorting order

File: core/src/main/scala/org/apache/spark/rdd/NewHadoopRDD.scala [hunk 7]
```diff
@@ -33,7 +33,7 @@ import org.apache.spark._
 import org.apache.spark.executor.DataReadMethod
 import org.apache.spark.mapreduce.SparkHadoopMapReduceUtil
 import org.apache.spark.rdd.NewHadoopRDD.NewHadoopMapPartitionsWithSplitRDD
-import org.apache.spark.util.{SerializableConfiguration, Utils}
+import org.apache.spark.util.{SerializableConfiguration, Utils, ShutdownHookManager}
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/rdd/NewHadoopRDD.scala]: nit: sorting order (well, in pretty much all imports you're adding)

File: core/src/main/scala/org/apache/spark/util/ShutdownHookManager.scala [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+
+import java.io.File
+import java.util.PriorityQueue
+
+import scala.util.{Failure, Success, Try}
+
+import org.apache.hadoop.fs.FileSystem
+import org.apache.spark.Logging
+
+import tachyon.client.TachyonFile
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/util/ShutdownHookManager.scala]: nit: this should come before `org.apache.spark` imports. (Just follow the pattern in other files.)

**Comment by vanzin** [comment]
LGTM, just have style nits left. I'll let this sit for a little bit to see if anyone else has comments.

File: core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala (hunk 6 above)
Comment (michellemay) [inline core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala]: Indeed, they didn't look like they were sorted in the first place..  So I tried to leave it that way everywhere.  But I agree with you that it's better sorted.

**Comment by vanzin** [comment]
Alright, merging. Will fix the imports in the process.

**Comment by vanzin** [comment]
Merged to master, branch-1.5 and branch-1.4. Thanks!

I fixed some minor conflicts in the 1.4 port (mostly import statements), made sure the build passed, and ran core/ unit tests locally.

**Comment by michellemay** [comment]
Awesome!  Thanks Marcelo for your implication!

**Pull request closed without merging**